### ISSUE TITLE
Clock Mesh Extension of CTS/PDN

### ISSUE THREAD

**Issue by QuantamHD** [The-OpenROAD-Project/OpenROAD#2202]
@arlpetergadfort Came across this [slide deck](https://people.ece.ubc.ca/ameer/publications/Abdelhadi-Talk-2010May-GLSVLSI2010-ClockMeshSynthesis.pdf) on the creation of Clock Meshes. Seems like it makes CTS much easier at the cost of higher power. 

Do you think you and I could extend PDN to generate the clock mesh, and then we could hook up the CTS to this new clock distribution network?

**Comment by maliberty** [comment]
It also takes up more routing resource but is a valid strategy for some cases.  There is a bunch of ongoing enhancement to the current CTS but I'm open to having further contributions so long as we coordinate to avoid collisions.  We are far from optimal in our current approach and I think a lot of benefit will be realized with the current work.

**Comment by mithro** [comment]
@QuantamHD - @mkkassem used a clock mesh on the metal layer to fix the hold issues on the ChipIgnite-1 silicon.

**Comment by mithro** [comment]
@mguthaus from OpenRAM also had a poster on clock meshes on his wall when I last visited.

**Comment by maliberty** [comment]
> @QuantamHD - @mkkassem used a clock mesh on the metal layer to fix the hold issues on the ChipIgnite-1 silicon.

I believe it was more of adding ad-hoc crosslinks in the clock tree rather than a true mesh.

**Comment by maliberty** [comment]
Note that delay calculation for meshes is much harder due to the multiple drivers and is not currently supported in sta.  With CCS it would become possible.

**Comment by mithro** [comment]
@maliberty - I believe @mkkassem just did spice simulation of the clocks for STA analysis?

**Comment by maliberty** [comment]
@mithro yes but you need to build a flow to back annotate those results into sta so you can do optimization/signoff.  That is the pre-CCS methodology.  If we want to make this generally available someone will need to automate that.

**Comment by QuantamHD** [comment]
I assumed we could link in ngspice to do that for us.

**Comment by mithro** [comment]
@maliberty - Is there a potential way to create a generic way to send things to a spice simulator and pull the timing information back into OpenROAD (even if extremely slow)?

**Comment by mithro** [comment]
Is there some relationship with BigSpicy? https://github.com/google/bigspicy

**Comment by arlpetergadfort** [comment]
@QuantamHD I see why you would want to use PDN for this, but I think it would be better to augment CTS to be able to build the grid itself and we can work to expose the DRC checking in PDN to be able to build vias and the grid correctly?

**Comment by maliberty** [comment]
Perhaps there is some code that could be an odb utility which both could share

**Comment by tspyrou** [comment]
@mithro I think that for a flow like this to work there would have to be a layer on top of the circuit simulator which sets up stage based runs. If the mesh is levelized from inputs to register leafs, each level can be run in parallel and in order with level 2 using slew waveforms from level one etc. From there an SDF can be generated which is fed into STA. This is a well known approach to clock mesh spice simulation which is scalable and accurate. These numbers can also be used as a baseline to measure STA accuracy when STA is enhanced to support meshes.

**Comment by maliberty** [comment]
This issue has been automatically marked as stale because it has not had recent activity. It will be closed in 21 days if no further activity occurs. Remove the Stale label or comment to keep it open.

**Comment by mguthaus** [comment]
We have a version of Clock Mesh Synthesis working and will eventually submit a PR. Closing now.

**Comment by npip99** [comment]
@mguthaus hey - do you have a current WIP commit for that PR? Happy to test it and see if it works for me

**Comment by mguthaus** [comment]
@npip99 We have the physical implementation but not the timing done. We have a few research papers on how to do the timing, but that isn't integrated into OpenROAD yet.

**Comment by npip99** [comment]
@mguthaus ok cool! I'd be happy to help collaborate on this if you could help point me to the research papers and/or partial implementation?
